Path/IO.hs

```diff
@@ -314,11 +329,187 @@ listDir path = do
+          case subdirs \\ xdirs of
+            [] -> return $ Just ()
+            ds -> runMaybeT $ mapM_ (MaybeT . walkAvoidLoop traversed) ds
+
+    checkLoop traversed dir = do
+      st <- liftIO $ getFileStatus (toFilePath dir)
+      let ufid = ((deviceID st), (fileID st))
+
+      -- check for loop, have we already traversed this dir?
+      if (Set.member ufid traversed)
+      then return Nothing
+      else return $ Just (Set.insert ufid traversed)
+
+-- | Create a descend handler for `walkDir` which excludes sub-directories
+-- matching a predicate, from descending.
+
+descendExcluding :: MonadIO m
+  => (Path Abs Dir -> m Bool)
+     -- ^ Directory match predicate
+  -> (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m WalkAction)
+     -- ^ Descend handler
+descendExcluding p = handler
+  where handler _ d _ = return . WalkExclude =<< filterM p d
+
+-- | Create a descend handler for `walkDir` which descends only sub-directories
+-- matching a predicate.
+
+descendWith :: (MonadIO m
+#if !MIN_VERSION_base(4,8,0)
+  , Functor m
+#endif
+  )
+  => (Path Abs Dir -> m Bool)
+     -- ^ Directory match predicate
+  -> (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m WalkAction)
+     -- ^ Descend handler
+descendWith p = handler
+  where handler paren d f = descendExcluding (fmap not . p) paren d f
+
+-- | Similar to 'walkDir' but accepts a 'Monoid' returning, output
+-- writer as well. Values returned by the output writer invocations are
+-- accumulated and returned.
+--
+-- Both, the descend handler as well as the output writer can be used for side
+-- effects but keep in mind that the output writer runs before the descend
+-- handler.
+
+walkDirAccum
+  :: (MonadIO m, MonadThrow m, Monoid o)
+  => Maybe (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m WalkAction)
+    -- ^ Descend handler (@dir -> subdirs -> files -> 'WalkAction'@),
+    -- descend the whole tree if omitted
+  -> (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File] -> m o)
+     -- ^ Output writer (@dir -> subdirs -> files -> o@)
+  -> Path b Dir
+     -- ^ Directory where traversal begins
+  -> m o
+     -- ^ Accumulation of outputs generated by the output writer invocations
+walkDirAccum dHandler writer topdir = do
+  ((), result) <- runWriterT $ walkDir handler topdir
+  return result
+  where
+    handler dir subdirs files = do
+      res <- lift $ writer dir subdirs files
+      tell res
+      case dHandler of
+        Just h -> lift $ h dir subdirs files
+        Nothing -> return (WalkExclude [])
+
+-- | Create an output writer for `walkDirAccum` which writes dirs and
+-- files matching corresponding predicates, as a tuple of lists.
+
+writeWith :: MonadIO m
+  => (Path Abs Dir -> m Bool)                 -- ^ Directory match predicate
+  -> (Path Abs File -> m Bool)                -- ^ File match predicate
+  -> (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File]
+      -> m ([Path Abs Dir], [Path Abs File])) -- ^ Output tuple writer
```

Comment:
I indented that because it is a continuation of a grouped type signature. The grouping may not be evident without the indentation.